### PR TITLE
Initial dot product attention

### PR DESCRIPTION
**Pull request by joaogui1**
Adds attention, per #61
So, first I'm really sorry about taking so long, but college got complicated in the pandemic and I wasted a lot of time getting organized.
Also, Attention is a quite general concept, and even implementations of the same type of attention differ significantly ([haiku](https://github.com/deepmind/dm-haiku/blob/a4b973b48e50db2309012e33f7033860f15d8ea0/examples/transformer/model.py), [flax](https://github.com/google/flax/blob/master/flax/linen/attention.py)) 
So @david-berthelot and @aterzis-google I would like to ask a few questions just to make sure my implementation is going in the right direction

1. I think I will implement a dot product attention, a multi-head attention and a masked attention, is that ok?
2. What do you think of the dot product attention implementation? What do you think I need to change?
Thanks for the patience and opportunity.

### CODE REVIEW

**Comment by david-berthelot** [comment]
Thanks, just a heads we'll get back to you sometime next week. I think what you plan to implement makes sense, but we'll discuss more in details next week (might bring in more reviewers for expertise on this particular topic).

**Review by ebrevdo — COMMENTED** [review]
1. This code needs a unit test.
2. Are you sure you want to use the same initializer for all weights?
3. Please add comments explaining what 'b', 't', 'h', 'd', and 'T' are for the einsums.

**Comment by joaogui1** [comment]
1. What kind of unit tests? Just comparing a couple inputs with the desired outputs?
2 and 3: Done

**Comment by joaogui1** [comment]
Hey @ebrevdo I made the implementation multi-head, added the option for masking, and made the einsum comments clearer. Now I think I just need to:

- [ ] Update the docstring
- [ ] Change the dot products to linear (I'm not sure I should do it, I've seen conflicting definitions)
- [ ] Add tests (can you answer my last question about how should I write the tests please)

Is that it?

File: objax/nn/layers.py [hunk 1]
```diff
@@ -248,36 +248,36 @@ def __init__(self, qin: int, kin: int, vin: int, dout: int, num_heads: int = 1,
...
         head_shape = (batch_size, sequence_length, self.num_heads, head_dim)
-        q = jn.reshape(jn.dot(x_q, self.w_q.value), head_shape)
-        k = jn.reshape(jn.dot(x_k, self.w_k.value), head_shape)
-        v = jn.reshape(jn.dot(x_v, self.w_v.value), head_shape)
+        q = jn.reshape(x_q.dot(self.w_q.value), head_shape)
+        k = jn.reshape(x_k.dot(self.w_k.value), head_shape)
+        v = jn.reshape(x_v.dot(self.w_v.value), head_shape)
 
-        # b: batch, t: query sequence lenght, T: key-value sequence length
+        # b: batch, t: query sequence length, T: key-value sequence length
```
Comment (ebrevdo) [inline objax/nn/layers.py:272]: no more batch, right?

File: objax/nn/layers.py [hunk 2]
```diff
@@ -248,36 +248,36 @@ def __init__(self, qin: int, kin: int, vin: int, dout: int, num_heads: int = 1,
...
-        scores = jn.einsum('bthd,bThd->bhtT', q, k) / jn.sqrt(head_dim)
+        scores = jn.einsum('thd,Thd->htT', q, k) / jn.sqrt(head_dim)
         if mask is not None:
             scores = scores * mask - 1e10 * (1 - mask)
         weights = functional.softmax(scores)
-        attention = jn.einsum('bhtT,bThd->bthd', weights, v)
-        attention = jn.reshape(attention, [batch_size, sequence_length, self.dout])
-        return jn.dot(attention, self.w_o.value)
+        attention = jn.einsum('htT,Thd->thd', weights, v)
+        attention = jn.reshape(attention, [batch_size, sequence_length, self.nout])
```
Comment (ebrevdo) [inline objax/nn/layers.py:279]: batch_size?

**Review by ebrevdo — CHANGES_REQUESTED** [review]
this still needs a unit test for inputs w/o batch dimensions.

@david-berthelot you may want to look at the other libraries' Attention impls and see if this API sufficiently matches, and give final approval.

**Comment by joaogui1** [comment]
> Change the dot products to linear (I'm not sure I should do it, I've seen conflicting definitions)
Can someone answer this?

**Comment by joaogui1** [comment]
> > Change the dot products to linear (I'm not sure I should do it, I've seen conflicting definitions)
> 
> Can someone answer this?

Pinging @ebrevdo @david-berthelot

**Comment by david-berthelot** [comment]
I don't really know the answer.

The way I typically solve such situations is by doing two versions: one using linear, one using einsum and then I can compare to make a call on which one we like best.

**Comment by joaogui1** [comment]
With einsum the code reads like this:
```Python
scores = jn.einsum('thd,Thd->htT', q, k) * functional.rsqrt(head_dim)
if mask is not None:
    scores = scores * mask - 1e10 * (1 - mask)
weights = functional.softmax(scores)
attention = jn.einsum('htT,Thd->thd', weights, v)
```

Without einsum this is how it will look like:
```Python
q = q.transpose(1, 2, 0)
k = k.transpose(1, 0, 2)
scores = q.dot(k) * functional.rsqrt(head_dim)
if mask is not None:
    scores = scores * mask - 1e10 * (1 - mask)
weights = functional.softmax(scores)
v = v.transpose(1, 0, 2)
attention = weights.dot(v)
attention = attention.transpose(1, 0, 2)
```

**Comment by joaogui1** [comment]
Though my question was not about einsum vs dot product, but rather between using a matrix multiplication to create the embeddings (x -> q, k, v) vs using a linear layer (matmul + bias) to accomplish it. When reading "Attention is All you Need" I had understood we used just a matmul, but in both implementations I listed in my first comment they used linear layers.

**Comment by david-berthelot** [comment]
`einsum` looks readable enough to me, concerning biases I don't have enough experience to answer properly.

@aterzis-google @ebrevdo Can you provide some feedback please?

**Comment by ebrevdo** [comment]
> `einsum` looks readable enough to me, concerning biases I don't have enough experience to answer properly.
> 
> @aterzis-google @ebrevdo Can you provide some feedback please?

I would follow the implementation in trax for this; trax was written by creators of transformer and their impl should be treated as ground truth.

https://github.com/google/trax/blob/master/trax/layers/attention.py

**Comment by joaogui1** [comment]
Ok folks, I was trying to finish this and realized I just don't know how (specially how to deal with tensor inputs to dense layers), I think I'm not as familiar with the library as I thought. I'm deeply sorry for wasting your time

**Comment by david-berthelot** [comment]
No worries, we'll still be happy to have your contributions for other tasks if you have the time and motivation to contribute!

**Pull request closed without merging**